<center>
    
#   Optimización para Machine learning
### Universidad de la Frontera
### Facultad de Ingeniería y ciencias aplicadas
### Minor de Análisis de Datos

#### Juan Aedo, Matías Palma, Gustavo Soto, José Bustamante, Lucas Williamson

#### Función 1
$$n=10, m=n+2 $$

$$f_k(w)=w_k-1,~~k=1,\cdots, n $$

$$f_{n+1}(w)=\sum_{k=1}^n k(w_k-1) $$

$$f_{n+2}(w)=\left(\sum_{k=1}^n k(w_k-1) \right)^2$$

#### Pregunta 1

1. Implemente el algoritmo de descenso del gradiente.

2. Utilice su implementacion del GD para encontrar un punto optimal de la primera función asignada a su grupo.

3. Compare los resultados utilizando distintos puntos iniciales, y estime cuántas iteraciones necesita el algorimto para mejorar su resultado en un factor de 0.1

In [1]:
n1 = 10
m1 = n1+2

In [2]:
import numpy as np

def f1(w,k):
    if len(w) != n1:
        return('Error de dimensión')
        
    if k > m1:
        return('Error de función, no existe f_'+str(k))
        
    if k < (n1+1):
        return (w[k-1]-1)
        
    elif k == n1+1:
        return (np.sum([i*(w[i-1]-1) for i in range(1,n1+1)]))
        
    elif k == n1+2:
        return (np.sum([(i*(w[i-1]-1))**2 for i in range(1,n1+1)]))

def F1(w):
    suma = sum([f1(w,k)**2 for k in range(1,m1+1)])
    return suma

In [3]:
w0=np.array([0,0,0,0,0,0,0,0,0,0])
len(w0)

10

In [4]:
n2 = 6
m2 = 10000

In [5]:
def f2(w,k):   
    if len(w) != n2:
        return('Error de dimensión')
    
    if k <= m2:
        t = k*0.1
        y = np.exp(-t)-5*np.exp(-10*t)+3*np.exp(-4*t)
        
        return(w[3] * np.exp(-t * w[0]) - w[3] * np.exp(-t * w[1]) + w[5] * np.exp(-t * w[4]) - y)

def F2(w):
    suma = sum([f2(w,k)**2 for k in range(1,m2+1)])
    return suma

In [6]:
w1 = [1,2,3,1,1,1,1,1,1,1]
w2 = [1,2,3,4,5,6]
len(w1),len(w2)

(10, 6)

In [7]:
F2(w2)

15.15470794940952

In [8]:
F1(w1)

1669

<center>
    
### Se calcula el gradiente de la siguiente forma
$$\nabla F_1(w)_k=2w_k-2+2k\sum_{i=1}^n({iw_{i-1}-1})+4k\left(\sum_{i=1}^n{iw_{i-1}-1}\right)^3 $$

In [17]:
def gradiente1(wk): 
    sum1=0
    grad=np.zeros(10)
    for i in range (0,len(wk)):
        sum1=sum1+(i+1)*(wk[i]-1)
    for i in range(0,len(wk)):
        grad[i]=(2*wk[i]-2)+2*(sum1)*(i+1)+4*((sum1)**3)*(i+1)
    return grad

In [55]:
#puntos inciales 
w0 = np.array([1,1,1,1,1,1,1,1,1,1])
w1=w0*0.9
w2=w0*1.1
w3=w0*0.95
w4 = 0*np.array([1,1,1,1,1,1,1,1,1,1])
w5= np.array([1,1,1,1,1,1,1,1,1,1])*2
len(w1)

10

In [19]:
gradiente1(w1)

array([ -676.7, -1353.2, -2029.7, -2706.2, -3382.7, -4059.2, -4735.7,
       -5412.2, -6088.7, -6765.2])


    
#### Método del descenso del Gradiente

Algoritmo en base a la iteración
$$w_{k+1}=w_k-\alpha\nabla F_1(w_k)$$

In [101]:

def metodo_grad(wk,a):
    cont=0
    gnorm=1
    epsilon=0.01
    while gnorm>epsilon:
        wk=wk-a*gradiente1(wk)
        gnorm=np.linalg.norm(gradiente1(wk))
        cont=cont+1
    print("Termino el metodo")
    print("En la iteracion = ",cont)
    print("Con funcion resultado", F1(wk))
    print("donde la norma del gradiente es", gnorm)
    print("con vector solucion ")
    print(wk)
    return wk
                

In [102]:
ws=metodo_grad(w1,0.00004)



Termino el metodo
En la iteracion =  42209
Con funcion resultado 2.5190920452182336e-05
donde la norma del gradiente es 0.009999472249660621
con vector solucion 
[0.99707245 0.99756038 0.9980483  0.99853623 0.99902415 0.99951208
 1.         1.00048792 1.00097585 1.00146377]


In [103]:
ws=metodo_grad(w2,0.00004)



Termino el metodo
En la iteracion =  42209
Con funcion resultado 2.5190920452167788e-05
donde la norma del gradiente es 0.009999472249657653
con vector solucion 
[1.00292755 1.00243962 1.0019517  1.00146377 1.00097585 1.00048792
 1.         0.99951208 0.99902415 0.99853623]


In [104]:
ws=metodo_grad(w3,0.00004)


Termino el metodo
En la iteracion =  42209
Con funcion resultado 2.5190920452182336e-05
donde la norma del gradiente es 0.009999472249660621
con vector solucion 
[0.99707245 0.99756038 0.9980483  0.99853623 0.99902415 0.99951208
 1.         1.00048792 1.00097585 1.00146377]


Curiosamente todas convergen en la misma iteracion, debe ser por que todas parten desde puntos similares.
Aunque la funcion resultado es del mismo orden de magnitud tienen algunos decimales de diferencia.
la norma del gradiente es bastante similar tambien

El método sólo funciona para puntos iniciales entre $0.9$ y $1.1$ ya que para otros puntos el gradiente se dispara a numeros muy grandes.   



In [105]:
ws=metodo_grad(w5,0.000000437)

Termino el metodo
En la iteracion =  10
Con funcion resultado nan
donde la norma del gradiente es nan
con vector solucion 
[nan nan nan nan nan nan nan nan nan nan]


C:\Users\Juanoooo\AppData\Local\Temp/ipykernel_15912/2543672446.py:7: RuntimeWarning: overflow encountered in double_scalars
  grad[i]=(2*wk[i]-2)+2*(sum1)*(i+1)+4*((sum1)**3)*(i+1)
C:\Users\Juanoooo\AppData\Local\Temp/ipykernel_15912/1041805322.py:6: RuntimeWarning: invalid value encountered in subtract
  wk=wk-a*gradiente1(wk)


In [106]:
ws=metodo_grad(w4,0.000000437)

Termino el metodo
En la iteracion =  10
Con funcion resultado nan
donde la norma del gradiente es nan
con vector solucion 
[nan nan nan nan nan nan nan nan nan nan]


C:\Users\Juanoooo\AppData\Local\Temp/ipykernel_15912/2543672446.py:7: RuntimeWarning: overflow encountered in double_scalars
  grad[i]=(2*wk[i]-2)+2*(sum1)*(i+1)+4*((sum1)**3)*(i+1)
C:\Users\Juanoooo\AppData\Local\Temp/ipykernel_15912/1041805322.py:6: RuntimeWarning: invalid value encountered in subtract
  wk=wk-a*gradiente1(wk)


Para puntos fuera del rango dicho anteriormente el gradiente se dispara a numeros muy grandes lo que resulta en errores de calculo.

Como el vector solucion obvio es [1,1,1,1,1,1,1,1,1,1] con solucion =0 esto tiene sentido

Para mejorar el resultado en un factor de 0.1 basta con el siguiente algoritmo

In [107]:
def metodo_gradR(wk,a):
    cont=0
    wsol=np.ones(10)
    epsilon=0.1
    gnorm=1
    while gnorm>epsilon :
        wk=wk-a*gradiente1(wk)
        gnorm=np.linalg.norm(wk-wsol)
        cont=cont+1
    print("Termino el metodo")
    print("En la iteracion = ",cont)
    print("donde la norma de wk-wsol=", gnorm,"<0.1")
    print("Con funcion resultado", F1(wk))
    print("con vector solucion ")
    print(wk)
    return wk

In [108]:
ws=metodo_gradR(w1,0.00004)


Termino el metodo
En la iteracion =  4764
donde la norma de wk-wsol= 0.09999347825370838 <0.1
Con funcion resultado 0.04096661577921959
con vector solucion 
[0.94144981 0.95120818 0.96096654 0.97072491 0.98048327 0.99024164
 1.         1.00975836 1.01951673 1.02927509]


Se necesitaron 4764 iteraciones para llegar a un factor de 0.1
